elements/lisk-chain/src/data_access/storage.ts

```diff
@@ -12,291 +12,537 @@
+			blocks.push({ ...header, transactions });
+		}
 
 		return blocks;
 	}
 
 	public async getLastBlock(): Promise<BlockJSON> {
-		const [lastBlock] = await this._storage.entities.Block.get(
-			{},
-			{ sort: 'height:desc', limit: 1, extended: true },
-		);
+		const header = await this.getLastBlockHeader();
+		const transactions = await this._getTransactions(header.id);
 
-		return lastBlock;
+		return {
+			...header,
+			transactions,
+		};
 	}
 
-	public async getTempBlocks(): Promise<TempBlock[]> {
-		const tempBlocks = await this._storage.entities.TempBlock.get(
-			{},
-			{ sort: 'height:asc', limit: null },
-		);
+	public async getTempBlocks(): Promise<BlockJSON[]> {
+		const stream = this._db.createReadStream({
+			gte: getFirstPrefix(DB_KEY_TEMPBLOCKS_HEIGHT),
+			lte: getLastPrefix(DB_KEY_TEMPBLOCKS_HEIGHT),
+			reverse: true,
+		});
+		const tempBlocks = await new Promise<BlockJSON[]>((resolve, reject) => {
+			const blocks: BlockJSON[] = [];
+			stream
+				.on('data', ({ value }) => {
+					blocks.push(value);
+				})
+				.on('error', error => {
+					reject(error);
+				})
+				.on('end', () => {
+					resolve(blocks);
+				});
+		});
 
 		return tempBlocks;
 	}
 
 	public async isTempBlockEmpty(): Promise<boolean> {
-		const isEmpty = await this._storage.entities.TempBlock.isEmpty();
+		const stream = this._db.createReadStream({
+			gte: getFirstPrefix(DB_KEY_TEMPBLOCKS_HEIGHT),
+			lte: getLastPrefix(DB_KEY_TEMPBLOCKS_HEIGHT),
+			limit: 1,
+		});
+		const tempBlocks = await new Promise<BlockJSON[]>((resolve, reject) => {
+			const blocks: BlockJSON[] = [];
+			stream
+				.on('data', ({ value }) => {
+					blocks.push(value);
+				})
+				.on('error', error => {
+					reject(error);
+				})
+				.on('end', () => {
+					resolve(blocks);
+				});
+		});
 
-		return isEmpty;
+		return tempBlocks.length === 0;
 	}
 
 	public async clearTempBlocks(): Promise<void> {
-		await this._storage.entities.TempBlock.truncate();
+		await this._db.clear({
+			gte: getFirstPrefix(DB_KEY_TEMPBLOCKS_HEIGHT),
+			lte: getLastPrefix(DB_KEY_TEMPBLOCKS_HEIGHT),
+		});
 	}
 
 	public async deleteBlocksWithHeightGreaterThan(
 		height: number,
 	): Promise<void> {
-		await this._storage.entities.Block.delete({
-			// eslint-disable-next-line camelcase
-			height_gt: height,
-		});
+		const lastBlockHeader = await this.getLastBlockHeader();
+		const batchSize = 5000;
+		const loops = Math.ceil((lastBlockHeader.height - height + 1) / batchSize);
+		const start = height + 1;
+		// tslint:disable-next-line no-let
+		for (let i = 0; i < loops; i += 1) {
+			// Get all the required info
+			const startHeight = i * batchSize + start + i;
+			const endHeight = startHeight + batchSize - 1;
+			const headers = await this.getBlockHeadersByHeightBetween(
+				startHeight,
+				endHeight,
+			);
+			const blockIDs = headers.map(header => header.id);
+			const transactionIDs = [];
+			const batch = this._db.batch();
+			for (const blockID of blockIDs) {
```

Comment:
If not possible to avoid the nested loop, can we move this to a helper function? (as if this code grows the nested loop makes it harder to read and reason about)

Reply:
As above, this function will be removed in https://github.com/LiskHQ/lisk-sdk/issues/5320